### PR TITLE
Add Memory Functions (MS, MR, M+, M- and MC) and Grand Total to extended functionalities

### LINKED ISSUE #145: HistoryDialog uses get_action_area
../src/HistoryDialog.vala:55.13-55.27: warning: Gtk.Dialog.get_action_area has been deprecated since 3.12

### PR DESCRIPTION
**Pull request by fabiozaramella**
As seen in most digital and real-life calculators (e.g. Windows 10 calculator).

### CODE REVIEW

File: src/MainWindow.vala
Comment (jeremypw) [inline src/MainWindow.vala:199]: The tooltip should indicate the stored value will overwrite any existing value (or maybe this button should be insensitive if the memory contains a value?) Perhaps "Set memory value" is clearer.  Maybe nice to actually include the value the memory will take.

As this is a destructive action maybe it should be styled as such?

File: src/MainWindow.vala
Comment (jeremypw) [inline src/MainWindow.vala:200]: "Use " is not really appropriate - all this button does is to write the value at the cursor position. "Recall" is maybe better. Maybe nice to include actual value?

File: src/MainWindow.vala
Comment (jeremypw) [inline src/MainWindow.vala:203]: "and replace it" is not really necessary.

File: src/MainWindow.vala
Comment (jeremypw) [inline src/MainWindow.vala:524]: For some reason, the entry cursor ends up on the wrong side (left) of grand total.

**Review by jeremypw — CHANGES_REQUESTED** [review]
See inline comments.

Another couple of issues I have noticed:
The recalled value can differ slightly from the one that was stored. Try storing  "1.1", clear entry and recall. You get "1.1000000000000001".  May have to remember the maximum precision of the numbers that were put into the memory and apply that when recalling.

It is not obvious what "Grand Total" does.  It adds all numbers in history but this list was, I think, intended more for undoing.  Also there is no way of seeing, editing or clearing the history (other than repeatedly pressing `<Ctrl>Z` or restarting the app.  I believe some physical calculators have a separate grand total memory which can be cleared by pressing the GT button twice?

**Comment by fabiozaramella** [comment]
> See inline comments.
> 
> Another couple of issues I have noticed:
> The recalled value can differ slightly from the one that was stored. Try storing "1.1", clear entry and recall. You get "1.1000000000000001". May have to remember the maximum precision of the numbers that were put into the memory and apply that when recalling.
> 
> It is not obvious what "Grand Total" does. It adds all numbers in history but this list was, I think, intended more for undoing. Also there is no way of seeing, editing or clearing the history (other than repeatedly pressing `<Ctrl>Z` or restarting the app. I believe some physical calculators have a separate grand total memory which can be cleared by pressing the GT button twice? Or maybe the "MC" button should clear the grand total memory as well?

Thanks for the reply! I'll try to address all these points asap. 😊 But I don't think MC should clear history, because on physical calculators it only clears the value([s] in some more advanced calculators like the one in Windows 10, which allow storing and recalling multiple values in memory similarly to our history dialog but i felt this would feel like a duplicate function and decided to keep it simple for the time being) in the memory not in the history. Also, usually GT just evaluate a total sum of all previous results that's why i picked them from the history. However, I agree that should be possible to remove (idk about edit, it doesn't really make sense in the context pf history dialog) unwanted entry from the list. 

About the MC button do you think it would make sense to set it sensitive when there isn't already a stored value and set it to unsensitive otherwise changing the tooltip to something like "A value is already stored in memory"? At least if we decide to keep it simple allowing one value at a time.

**Comment by jeremypw** [comment]
I am happy to merge a simple implementation, subject to design approval, so long as it behaves more or less in line with most physical and digital calculators.  We can always iterate it later if required to improve the grand total functionality.

> About the MC button do you think it would make sense to set it to sensitive when there isn't already a stored value and set it to insensitive otherwise changing the tooltip to something like "A value is already stored in memory"?

Do you mean the MS button? What you suggest would prevent a user from accidentally overwriting a memory value but it also kind of makes the MS button redundant since you can have the same effect by pressing MC and M+.  The space could be used for e.g. a clear history button.

**Comment by jeremypw** [comment]
As far as destructive styling is concerned I envisaged only having the text in the destructive color, not the whole button. That can be left to the design team.

**Comment by fabiozaramella** [comment]
> Do you mean the MS button? What you suggest would prevent a user from accidentally overwriting a memory value but it also kind of makes the MS button redundant since you can have the same effect by pressing MC and M+. The space could be used for e.g. a clear history button. Or the function `n!` has been requested.

Yes, i meant the MS button. I know it would be a little redundant but that's what calculators usually provide so idk.

Just updated the PR with the rewording you suggested. Still have to fix the correct value being stored and recalled.

**Comment by jeremypw** [comment]
Because it is not possible to store some decimal numbers in a finite number of binary digits the least significant decimal digits maybe inaccurate.  Double precision variables store 53 binary bits of precision which is equivalent to between 15 and 17 decimal digits (depending on the number). The simplest solution would be to simply truncate all decimal numbers to a maximum of 15 (or fewer) significant digits - which should be more than enough for most practical purposes.

**Comment by jeremypw** [comment]
@fabiozaramella I notice this PR has not had activity for a while and now has conflicts - is this something you are still available to work on?

**Comment by jeremypw** [comment]
Closing due to lack of response from author.

**Comment by fabiozaramella** [comment]
@jeremypw I'm sorry I didn't answer, I've been really busy in the last months. I'll try to get back to it if I can.

**Comment by fabiozaramella** [comment]
@jeremypw It should work as expected now, I'm waiting for your approval 😊

**Comment by fabiozaramella** [comment]
> See inline comments.
> 
> Another couple of issues I have noticed:
> The recalled value can differ slightly from the one that was stored. Try storing "1.1", clear entry and recall. You get "1.1000000000000001". May have to remember the maximum precision of the numbers that were put into the memory and apply that when recalling.
> 
> It is not obvious what "Grand Total" does. It adds all numbers in history but this list was, I think, intended more for undoing. Also there is no way of seeing, editing or clearing the history (other than repeatedly pressing `<Ctrl>Z` or restarting the app. I believe some physical calculators have a separate grand total memory which can be cleared by pressing the GT button twice? Or maybe the "MC" button should clear the grand total memory as well?

About the history... I think there should be a clear history button in the history dialog, maybe with destructive stile. In any case it should be addressed in another PR.

**Review by jeremypw — APPROVED** [review]
The most serious issue is now fixed - thanks!

The other issues are subject to discussion so can be left to future PRs.

**Pull request merged**
